In [1]:
!pip install pymongo

# Connecting to mongodb


In [1]:
# importing the required libraries
import pymongo
import pprint
import json
import warnings
import pandas as pd, numpy as np
warnings.filterwarnings('ignore')

# connect to the mongoclient
client = pymongo.MongoClient('mongodb://localhost:27017')

# get the database
database = client['intern']

In [2]:
sensor_data = database.get_collection("sensor_data")

In [3]:
column_names = sensor_data.distinct("sensor_name")

### Single feature analysis for data from mongo end

In [5]:
motor_voltage = sensor_data.find({"sensor_name":'Motor Voltage'}, {"data":1})
for doc in motor_voltage:
    print(len(doc['data']))

5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760
5760


#### each unit of list in data has two set of variables one is the timestamp and the other is that specifc sensot values

## Initial analysis to check how to combine columns

In [9]:
cursor = sensor_data.find({"sensor_name":'Motor Voltage'}, 
                                    {"data":1})

pd_df = pd.DataFrame(columns=["timestamp",'value'])

for doc in cursor:
    temp_df = pd.DataFrame(doc['data'], columns=["timestamp",'value'])
    pd_df = pd.concat([pd_df, temp_df])

pd_df.sort_values(by='timestamp').reset_index(drop=True)

timestamp        value
0        2018-01-01 00:00:15  4234.068848
1        2018-01-01 00:00:30       4235.0
2        2018-01-01 00:00:45   4236.45459
3        2018-01-01 00:01:00  4232.772461
4        2018-01-01 00:01:15  4240.088379
...                      ...          ...
4204794  2019-12-31 23:58:45  4238.430176
4204795  2019-12-31 23:59:00  4232.068848
4204796  2019-12-31 23:59:15  4240.943848
4204797  2019-12-31 23:59:30  4235.980957
4204798  2019-12-31 23:59:45   4238.92627

[4204799 rows x 2 columns]

In [11]:
del pd_df
gc.collect()

226

In [6]:
pd.DataFrame(sensor_data.find_one({"sensor_name":'Motor Voltage'}, 
                                    {"data":1})["data"], 
                                    columns=["timestamp",'value']).merge(pd.DataFrame(sensor_data.find_one({"sensor_name":'Pump Shaft Speed'}, {"data":1})["data"], 
                                                                                        columns=['timestamp','value']), on='timestamp')

timestamp      value_x      value_y
0    2018-01-01 00:00:15  4234.068848  3589.281494
1    2018-01-01 00:00:30  4235.000000  3589.281982
2    2018-01-01 00:00:45  4236.454590  3589.282715
3    2018-01-01 00:01:00  4232.772461  3589.283203
4    2018-01-01 00:01:15  4240.088379  3589.283691
...                  ...          ...          ...
5755 2018-01-01 23:59:00  4233.307617  3589.990234
5756 2018-01-01 23:59:15  4242.801270  3589.990234
5757 2018-01-01 23:59:30  4229.829590  3589.990234
5758 2018-01-01 23:59:45  4234.789551  3589.990234
5759 2018-01-02 00:00:00  4236.454590  3589.990234

[5760 rows x 3 columns]

All the 5760 rows were displayed after merging two features, so no data loss is observed.

Columns names have to be changed to what they represent, some tranformation is required.

## Final data creation for modelling

### Extract timestamp as a single dataframe to merge to other columns

In [12]:
cursor = sensor_data.find({"sensor_name":'Motor Voltage'}, 
                                    {"data":1})

pd_df = pd.DataFrame(columns=["timestamp"])

for doc in cursor:
    temp_df = pd.DataFrame(doc['data'], columns=["timestamp",'value'])[['timestamp']]
    pd_df = pd.concat([pd_df, temp_df])

pd_df.sort_values(by='timestamp').reset_index(drop=True)

pd_df.head()

timestamp
0  2018-01-01 00:00:15
1  2018-01-01 00:00:30
2  2018-01-01 00:00:45
3  2018-01-01 00:01:00
4  2018-01-01 00:01:15

In [13]:
pd_df.shape

(4204799, 1)

### Created a single column df with just timestamp so that other columns can be merged to it one by one and we don't have to fruther clean it

In [34]:
len(column_names)

26

### As number of columns are only 26 no loop opimization is required, instead of using map or vectorization on python lists, as simple for loop will do ther trick

In [14]:
for col in column_names:
    cursor = sensor_data.find({"sensor_name":col}, 
                                    {"data":1})

    col_df = pd.DataFrame(columns=["timestamp",'value'])

    for doc in cursor:
        temp_df = pd.DataFrame(doc['data'], columns=["timestamp",'value'])
        col_df = pd.concat([col_df, temp_df])

    col_df = col_df.sort_values(by='timestamp').reset_index(drop=True).rename(columns ={"value":col})

    
                                    
    pd_df = pd_df.merge(col_df, on='timestamp')

pd_df

timestamp  \
0        2018-01-01 00:00:15   
1        2018-01-01 00:00:30   
2        2018-01-01 00:00:45   
3        2018-01-01 00:01:00   
4        2018-01-01 00:01:15   
...                      ...   
4204794  2019-12-31 23:58:45   
4204795  2019-12-31 23:59:00   
4204796  2019-12-31 23:59:15   
4204797  2019-12-31 23:59:30   
4204798  2019-12-31 23:59:45   

        Auxiliary Boilers A/B Feed Water Header Pressure 2  \
0                                              1305.491089   
1                                              1301.227783   
2                                              1296.178589   
3                                              1297.628296   
4                                              1299.078003   
...                                                    ...   
4204794                                        1305.224976   
4204795                                        1304.164673   
4204796                                        1308.058838   
4204797                                        1315.320679   
4204798                                         1313.70752   

        Auxiliary Boilers Feed Water Header Pressure  \
0                                        1305.893799   
1                                        1299.925171   
2                                        1294.646606   
3                                         1299.98877   
4                                        1300.749023   
...                                              ...   
4204794                                  1305.452515   
4204795                                  1303.207764   
4204796                                  1306.166016   
4204797                                  1312.660889   
4204798                                  1315.073364   

        Heat Recovery System Header Mass Flow  \
0                                 301258.6563   
1                                   303027.75   
2                                 301878.9688   
3                                 302023.0625   
4                                 301972.0625   
...                                       ...   
4204794                           325015.4688   
4204795                            324984.625   
4204796                             324521.25   
4204797                           324716.9688   
4204798                            324912.125   

        Heat Recovery System Header Mass Flow.1  \
0                                   304479.0625   
1                                   303782.2813   
2                                   303518.4375   
3                                    303576.375   
4                                   303999.5313   
...                                         ...   
4204794                             330229.3438   
4204795                             329936.0625   
4204796                             329425.6875   
4204797                               330124.75   
4204798                             330279.7813   

        Heat Recovery System Header Pressure  \
0                                1779.438599   
1                                1779.364014   
2                                1779.289551   
3                                1779.214966   
4                                1779.140381   
...                                      ...   
4204794                          1702.869385   
4204795                          1703.680542   
4204796                          1704.491699   
4204797                          1705.302856   
4204798                          1706.114014   

        Heat Recovery System Header Pressure.1  \
0                                  1782.647461   
1                                  1782.624756   
2                                  1782.602173   
3                                  1782.579468   
4                                  1782.556885   
...                                        ...   
4204794                            1706.575195   
4204795                            17

In [15]:
gc.collect()

50

In [16]:
pd_df.to_csv('data/sensor_data.csv')

NameError: name 'pd_df' is not defined